sycl/doc/extensions/AsyncGroupCopy/async_group_copy.asciidoc

```diff
@@ -0,0 +1,160 @@
+
+IMPORTANT: This specification is a draft.
+
+NOTE: The APIs described in this specification are experimental. Future versions of this extension may change these APIs in ways that are incompatible with the versions described here.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension is written against the SYCL 2020 revision 4 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+## Contributors
+
+* Gordon Brown
+* Tadej Ciglaric
+* Jack Kirk
+* Finlay Marno
+
+## Introduction
+
+This document proposes `async_group_copy`, which will generalize
+`async_work_group_copy` to work with sub-groups as well as work-groups.
+A `wait_for` function will also be also be added to block until the memory
+copy is complete and to ensure memory consistency for the group, this will use
+the same groups as `async_group_copy`.
+
+## Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros". Therefore, an implementation
+supporting this extension must predefine the macro
+SYCL_EXT_ONEAPI_ASYNC_GROUP_COPY to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro’s value
+to determine which of the extension’s APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension implementation.
+|===
+
+
+## New `async_group_copy` function.
+`async_group_copy` will work very similarly to `async_work_group_copy`, but
+with the ability to utilize individual sub-groups. The group will be provided
+as a template parameter, in the same fashion as `group_barrier`.
+`async_group_copy` is intended as a potential optimization, and may add more
+overhead than a simple synchronous implementation in some cases.
+
+`async_group_copy` will asynchronously copy a given number of values of type
+`dataT` from global memory to local memory or from local memory to global
+memory. If an asynchronous copy is not supported then `async_group_copy` will
+fall back on a synchronous copy. `async_group_copy` is a free function with the
+following properties:
+
+* The work-items will not necessarily be synchronized, so must have a
+  consistent view of the memory before the invocation.
+* The order of operations is not specified.
+* All work-items of the group are required to call the function with the same
+  arguments, otherwise the behaviour is undefined.
+* All work-items of the group are required to call the function in convergent
+  control flow, otherwise the behaviour is undefined.
+* The type of `dataT` should be a scalar or vector type.
```

Comment:
```suggestion
* The type of `dataT` must be a scalar or vector type.
```

If this is a requirement, use "must".

Also, is this requirement really what we want?  The term "scalar type" refers to any of the types listed in [4.14.1. "Scalar data types"](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_scalar_data_types), which seems reasonable.  However, "vector type" means either the `sycl::vec` type or the `sycl::marray` type as described in [4.14.2. Vector types](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:vector.type).

Note that the element type of `std::marray` is defined as "must be a numeric type as it is defined by C++ standard", which refers to the [C++ named requirement NumericType](https://en.cppreference.com/w/cpp/named_req/NumericType).  However, that requirement is very broad, allowing non-scalar types.

Therefore, it seems weird that the type passed to `async_group_copy` cannot be a non-scalar type, but it *can* be an `std::marray` of non-scalar types.  I presume this is not what we intend.

Maybe we want to say that the type passed to `async_group_copy` can be any trivially copyable type?

Reply:
I didn't realise that std::marray was for numeric types, that is a bit of a surprise to me. I've gone for trivially copyable as suggested.

Also I've tried to removed any weak language.

Reply:
@gmlueck Hi I've been looking into implementing this and I think it'll be easier for now to just keep it to scalar types and sycl::vec.
When implementing for trivially copyable types my plan was to cast the data to a char* but then it was impossible to do the stride correctly without know the size of the object, which would require a new spirv instruction.
Hopefully this still leaves scope for that in the future if it would be beneficial.

Reply:
@gmlueck @Pennycook 
I was talking to @AerialMantis earlier and we decided that we would actually like to keep trivially copyable in the proposal, but for the first implementation we will just implement SYCL scalar types and sycl::vec. In the future we would like to add a spirv extension much like OpGroupAsyncCopy, but also with a object shape (size in bytes) that will allow for the stride to be done properly.
Sorry about the flip-flop.

Reply:
In general, I think the specifications for DPC++ extensions should match our intended implementation in DPC++.  Therefore, if DPC++ intends to implement only scalar types and `sycl::vec` here, I think the spec should state those restrictions.  I agree that it would be nicer if the spec allowed any trivially copyable type.  If we decide to add that support later, we can update the specification to "revision 2" which broadens the set of allowable types.

The reason I prefer this approach is because DPC++ users can then rely on the extension specification as the documentation of the feature.  By contrast, if the specification is "aspirational", then we need some separate document describing which parts of the extension are really implemented.  This is less friendly to users.

I would consider an exception to this rule if there are other vendors who plan to implement this extension.  If the extension is DPC++ only for now, though, then I don't see any value in specifying some behavior that we don't implement in DPC++.

If you want to capture this issue in the spec, you could add a non-normative section at the end stating that we would like to broaden the set of allowed types in the future.  This section could even discuss the challenges with allowing trivially copyable types, noting that a SPIR-V extension is required.

Reply:
I've checked a bit more thoroughly this time by implementing the synchronous fall back that @Pennycook suggested for trivially copyable types, so I'm happy this it is actually possible.
I've also added a bit at the end describing the difficulties of OpGroupAsyncCopy for future reference.